src/relax/transform/split_calltir_by_pattern.cc

```diff
@@ -0,0 +1,745 @@
+#include <tvm/relax/expr_functor.h>
+#include <tvm/relax/transform.h>
+#include <tvm/relax/type.h>
+#include <tvm/tir/builtin.h>
+#include <tvm/tir/op.h>
+#include <tvm/tir/stmt_functor.h>
+
+#include "../../tir/schedule/ir_comparator.h"
+
+namespace tvm {
+
+const static constexpr char* kCutlassKernel = "cutlass_kernel";
+const static constexpr char* kCutlassCodegen = "cutlass_codegen";
+const static constexpr char* kCSource = "c_source";
+const static constexpr char* kCSourceFmt = "c_source_fmt";
+const static constexpr char* kCSourceFmtCuda = "cu";
+
+namespace tir {
+
+class BlockMatcher : public TensorizeComparator {
+ public:
+  using SymbolMap = std::unordered_map<Var, PrimExpr, ObjectPtrHash, ObjectPtrEqual>;
+  explicit BlockMatcher(const tir::PrimFunc& pattern, const Array<Var>& pattern_vars)
+      : TensorizeComparator(IRModule({{GlobalVar(""), pattern}}), false), pattern_(pattern) {
+    for (const auto& pattern_var : pattern_vars) {
+      this->pattern_vars_.insert(pattern_var);
+    }
+    this->evaluated_symbols.push_back(SymbolMap());
+  }
+
+  bool Match(const For& top) {
+    const ForNode* pattern_top = pattern_->body.as<BlockRealizeNode>()->block->body.as<ForNode>();
+    ICHECK(pattern_top) << "Invalid pattern function";
+    if (!VisitStmt(top, GetRef<Stmt>(pattern_top))){
+      return false;
+    }
+    // Get evaluated symbols, buffers from the pattern.
+    for (const auto& arg : pattern_->params) {
+      auto it = pattern_->buffer_map.find(arg);
+      if (it != pattern_->buffer_map.end()) {
+        auto itt = rhs_buffer_map_.find((*it).second);
+        ICHECK(itt != rhs_buffer_map_.end());
+        evaluated_buffers.push_back(itt->second);
+      }
+    }
+    return true;
+  }
+
+  std::vector<SymbolMap> evaluated_symbols;
+  std::vector<Buffer> evaluated_buffers;
+
+ private:
+  Optional<PrimExpr> QueryEvaluatedSymbols(const Var& var) {
+    for (const SymbolMap& symbol_map : evaluated_symbols) {
+      auto it = symbol_map.find(var);
+      if (it != symbol_map.end()) {
+        return it->second;
+      }
+    }
+    return NullOpt;
+  }
+
+  bool VisitExpr(const PrimExpr& lhs, const PrimExpr& rhs) final {
+    if (const auto* op = rhs.as<VarNode>()) {
+      if (pattern_vars_.count(GetRef<Var>(op))) {
+        // special case for pattern vars
+        const auto* lhs_ptr = lhs.as<VarNode>();
+        if (lhs_ptr == nullptr) {
+          if (lhs->IsInstance<tir::IntImmNode>() || lhs->IsInstance<tir::FloatImmNode>()) {
+            Optional<PrimExpr> value = QueryEvaluatedSymbols(GetRef<Var>(op));
+            if (value.defined()) {
+              if (!analyzer_.CanProveEqual(lhs, value.value())) return false;
+            } else {
+              evaluated_symbols.back()[GetRef<Var>(op)] = lhs;
+            }
+            return true;
+          } else {
+            return false;
+          }
+        }
+      }
+    }
+    // pattern_var * expr
+    if (const auto* rhs_ptr = rhs.as<MulNode>()) {
```

Comment:
consider reuse tools in https://github.com/tlc-pack/relax/blob/relax/src/arith/pattern_match.h